code/code/misc/offense.cc

```diff
@@ -766,332 +766,283 @@ static bool canFleeThisWay(TBeing *ch, dirTypeT dir)
-            FALSE, this, NULL, NULL, TO_CHAR, ANSI_RED);
-    return FALSE;
+    act("$n tries to flee but a strange power prevents $m from escaping.", false, this, nullptr,
+        nullptr, TO_ROOM);
+    act("A strange power prevents you from escaping.", false, this, nullptr, nullptr, TO_CHAR,
+        ANSI_RED);
+    return false;
   }
-  if (riding) {
-    // Let's make retreat skill have some chance of keeping you on the mount - Brutius 07/26/1999
-    if (!(doesKnowSkill(skill) && doesKnowSkill(skill2)) || 
-	!(bSuccess(skill) && bSuccess(skill2))) { 
+
+  /* 
+  Do skill checks now and use results throughout so bSuccess only gets 
+  called once per flee attempt. 
+  */
+  auto retreatSkill = getSkillNum(SKILL_RETREAT);  
+  auto doesKnowRetreat = doesKnowSkill(retreatSkill);
+  auto wasRetreatSuccessful = doesKnowRetreat && bSuccess(retreatSkill);  
+
+  auto panic = false;
+  auto rc = 0;
+  auto riderAsTBeing = dynamic_cast<TBeing *>(rider);
+
+  if (riding) {   
+    // Allow successful retreat and ride checks to avoid falling off mount
+    if (!doesKnowSkill(SKILL_RIDE) || !bSuccess(SKILL_RIDE) || !wasRetreatSuccessful) {
       sendTo("Your panic causes you to fall.\n\r");
       rc = fallOffMount(riding, POSITION_SITTING);
-      panic = TRUE;
-      if (IS_SET_DELETE(rc, DELETE_THIS)) 
-        return DELETE_THIS;
+      if (IS_SET_DELETE(rc, DELETE_THIS)) return DELETE_THIS;
+      panic = true;
     }
   } else if (getPosition() <= POSITION_SITTING) {
     addToMove(-10);
-    act("$n scrambles madly to $s feet!", TRUE, this, 0, 0, TO_ROOM);
-    act("Panic-stricken, you scramble to your feet.", TRUE, this, 0, 0, TO_CHAR);
+    // If retreat succeeded, simply being on the ground when fleeing doesn't
+    // result in panic. And you get a different message.
+    if (wasRetreatSuccessful) {
+      act("$n rolls aside and regains $s footing!", true, this, nullptr, nullptr, TO_ROOM);
+      act("You quickly roll aside and regain your footing.", true, this, nullptr, nullptr, TO_CHAR);
+    } else {
+      act("$n scrambles madly to $s feet!", true, this, nullptr, nullptr, TO_ROOM);
+      act("Panic-stricken, you scramble to your feet.", true, this, nullptr, nullptr, TO_CHAR);
+    }
+
     doStand();
-    addToWait(combatRound(1));
 
-    panic = TRUE;
-  } else if ((t = rider)) {
-    t->sendTo("Your mount panics and attempts to flee.\n\r");
-    rc = t->fallOffMount(this, POSITION_SITTING);
-    panic = TRUE;
+    // If retreat didn't succeed or char doesn't have it, add some lag
+    // and panic.
+    if (!wasRetreatSuccessful) {
+      addToWait(combatRound(1));
+      panic = true;
+    }
+    // If <this> is a mount being ridden
+  } else if (riderAsTBeing) {    
+    act("Your $O panics and attempts to flee!", true, riderAsTBeing, nullptr, this, TO_CHAR);
+
+    // Allow deikhans to pass checks and prevent getting bucked
+    if (
+      riderAsTBeing->doesKnowSkill(SKILL_CHIVALRY) &&
```

Comment:
Gameplay tweak/question - but would it make sense to also include any advanced riding and/or mount-specific mastery skill in the calculation?  We're bucking off the rider if they fail basic riding and chivalry checks, but it seems to me that someone who knows not just riding and chivalry but also has mastered ride mount_type (e.g., ride domestic) as well as advanced riding should get some bonuses to these checks.

Reply:
Definitely makes sense. I don't know how common being bucked off actually is, so initially I was just thinking "don't let Deikhans be easily bucked off because that sucks for gameplay" but it probably should be more specific to the type of mount. I'll look into it.

Reply:
This now instead relies on passing bSuccess checks for both the advanced riding skill for the type of mount being ridden and the calm mount skill. See https://github.com/sneezymud/sneezymud/pull/373/commits/9eafa770ec48ccd42c3287b72176a0950acf293e.